source/common/matcher/matcher.h

```diff
@@ -0,0 +1,167 @@
+    return evaluateMatch(*result.on_match_->matcher_, data);
+  }
+
+  return MaybeMatchResult{result.on_match_->action_cb_(), true};
+}
+
+/**
+ * Recursively constructs a MatchTree from a protobuf configuration.
+ */
+template <class DataType> class MatchTreeFactory {
+public:
+  MatchTreeFactory(ProtobufMessage::ValidationVisitor& validation_visitor)
+      : validation_visitor_(validation_visitor) {}
+
+  MatchTreeSharedPtr<DataType> create(const envoy::config::common::matcher::v3::Matcher& config) {
+    if (config.has_matcher_tree()) {
+      return createTreeMatcher(config);
+    } else if (config.has_matcher_list()) {
+      return createListMatcher(config);
+    } else {
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+
+private:
+  MatchTreeSharedPtr<DataType>
+  createListMatcher(const envoy::config::common::matcher::v3::Matcher& config) {
+    auto list_matcher =
+        std::make_shared<ListMatcher<DataType>>(createOnMatch(config.on_no_match()));
+
+    for (const auto& matcher : config.matcher_list().matchers()) {
+      list_matcher->addMatcher(createFieldMatcher(matcher.predicate()),
+                               *createOnMatch(matcher.on_match()));
+    }
+
+    return list_matcher;
+  }
+
+  FieldMatcherPtr<DataType> createFieldMatcher(
+      const envoy::config::common::matcher::v3::Matcher::MatcherList::Predicate& field_predicate) {
+    switch (field_predicate.match_type_case()) {
+    case (envoy::config::common::matcher::v3::Matcher::MatcherList::Predicate::kSinglePredicate):
+
+      return std::make_unique<SingleFieldMatcher<DataType>>(
+          createDataInput(field_predicate.single_predicate().input()),
+          createInputMatcher(field_predicate.single_predicate()));
+    case (envoy::config::common::matcher::v3::Matcher::MatcherList::Predicate::kOrMatcher): {
+      std::vector<FieldMatcherPtr<DataType>> sub_matchers;
+      for (const auto& predicate : field_predicate.or_matcher().predicate()) {
+        sub_matchers.emplace_back(createFieldMatcher(predicate));
+      }
+
+      return std::make_unique<AnyFieldMatcher<DataType>>(std::move(sub_matchers));
+    }
+    case (envoy::config::common::matcher::v3::Matcher::MatcherList::Predicate::kAndMatcher): {
+      std::vector<FieldMatcherPtr<DataType>> sub_matchers;
+      for (const auto& predicate : field_predicate.and_matcher().predicate()) {
+        sub_matchers.emplace_back(createFieldMatcher(predicate));
+      }
+
+      return std::make_unique<AllFieldMatcher<DataType>>(std::move(sub_matchers));
+    }
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+
+  MatchTreeSharedPtr<DataType>
+  createTreeMatcher(const envoy::config::common::matcher::v3::Matcher& matcher) {
+    auto multimap_matcher = std::make_shared<ExactMapMatcher<DataType>>(
+        createDataInput(matcher.matcher_tree().input()), createOnMatch(matcher.on_no_match()));
+
+    for (const auto& children : matcher.matcher_tree().exact_match_map().map()) {
```

Comment:
Missing branching/TODO for `prefix_match_map` and `custom_match`?

Reply:
Yep, added with NOT_IMPLEMENTED